### ISSUE TITLE
Attached role policcies are detected as "changed outside Terraform"  and added as "managed_policy_arns" attribute.

### ISSUE THREAD

**Issue by henninge** [hashicorp/terraform-provider-aws#20984]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version

    Terraform v1.0.5
    on darwin_amd64
    + provider registry.terraform.io/hashicorp/aws v3.59.0

### Affected Resource(s)

* aws_iam_role
* aws_iam_role_policy_attachment

### Terraform Configuration Files

```hcl
terraform {
  required_version = "~> 1.0"

  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 3.0"
    }
  }
}

data "aws_caller_identity" "current" {}

data "aws_iam_policy_document" "trust" {
  statement {
    effect = "Allow"

    principals {
      type        = "AWS"
      identifiers = [data.aws_caller_identity.current.arn]
    }

    actions = ["sts:AssumeRole"]
  }
}

resource "aws_iam_role" "testing_foo" {
  assume_role_policy = data.aws_iam_policy_document.trust.json
}

data "aws_iam_policy_document" "testing_foo" {
  statement {
    effect = "Allow"

    actions = [
      "s3:*",
    ]
    resources = ["*"]
  }
}

resource "aws_iam_policy" "testing_foo" {
  policy = data.aws_iam_policy_document.testing_foo.json
}

resource "aws_iam_role_policy_attachment" "testing_foo" {
  role       = aws_iam_role.testing_foo.name
  policy_arn = aws_iam_policy.testing_foo.arn
}
```

### Expected Behavior

As the documentation for `aws_iam_role` suggests this config should produce a consistent state as it uses `aws_iam_role_policy_attachment` but not `aws_iam_role.managed_policy_arns`.  No `managed_policy_arns`attritbute should be added to the role.

### Actual Behavior

`terraform apply` works as expected but a subsequent `terraform plan` detects the newly created attachments as "changed outside terraform" and plans to add a `managed_policy_arns` attribute to the `aws_iam_role` resource in the statate. This is done with `terraform apply -refresh-only` although this is forbidden by the documenation (either attribute or attachement resource, not both).

```
Note: Objects have changed outside of Terraform

Terraform detected the following changes made outside of Terraform since the last "terraform apply":

  # aws_iam_policy.testing_foo has been changed
  ~ resource "aws_iam_policy" "testing_foo" {
        id        = "arn:aws:iam::060779697980:policy/terraform-2021092200000000000000"
        name      = "terraform-2021092200000000000000"
      + tags      = {}
        # (5 unchanged attributes hidden)
    }
  # aws_iam_role.testing_foo has been changed
  ~ resource "aws_iam_role" "testing_foo" {
        id                    = "terraform-2021092200000000000000"
      ~ managed_policy_arns   = [
          + "arn:aws:iam::060779697980:policy/terraform-2021092200000000000000",
        ]
        name                  = "terraform-2021092200000000000000"
      + tags                  = {}
        # (9 unchanged attributes hidden)

        # (1 unchanged block hidden)
```

Removing the `aws_iam_role_policy_attachment` resource from the config results in the same two-step apply.

### Steps to Reproduce

1. `terraform apply`
2. `terraform plan`
3. `terraform apply -refresh-only`

### Important Factoids

I stumbled onto this during upgrade of the provider which is extra-confusing for the use since neither the config nor the deployed resources had changed.

This seems related to #20107  were people in the comments reported the same behavior for the `inline_policy` attribute.

### References

* #20107

**Comment by breathingdust** [comment]
Likely related: #19727

**Comment by benfain** [comment]
This is still occurring for me on TF v1.1.6 and AWS provider v3.74.3, @YakDriver should this be tracked in https://github.com/hashicorp/terraform-provider-aws/issues/23288 maybe? This is one is confusing for me because it's not like whitespace drift, the state would store a conflicting value. The only workaround I suppose is removing `aws_iam_role_policy_attachment` resources

**Comment by lblazewski** [comment]
Also seeing this for example in the https://github.com/terraform-aws-modules/terraform-aws-iam/tree/v5.10.0/modules/iam-role-for-service-accounts-eks module.

**Comment by jar-b** [comment]
In attempting to reproduce this issue, it appears to have been resolved by an upstream Terraform core change in [`v1.2.0`](https://github.com/hashicorp/terraform/blob/v1.2/CHANGELOG.md#120-may-18-2022) (specifically https://github.com/hashicorp/terraform/pull/30486):

> When showing a plan, Terraform CLI will now only show "Changes outside of Terraform" if they relate to resources and resource attributes that contributed to the changes Terraform is proposing to make.

The complete reproduction is documented below. We are proposing this issue be closed via the upstream fix, but will leave it open for now to allow time for responses.

---

### Setup

The following configuration was used to reproduce this issue.

<details>
<summary>Show Config</summary>

```hcl
terraform {
  required_version = "~> 1.0"

  required_providers {
    aws = {
      source  = "hashicorp/aws"
      #version = "~> 3.0"
      version = "3.59.0"
    }
  }
}

provider "aws" {}

data "aws_caller_identity" "current" {}

data "aws_iam_policy_document" "trust" {
  statement {
    effect = "Allow"

    principals {
      type        = "AWS"
      identifiers = [data.aws_caller_identity.current.arn]
    }

    actions = ["sts:AssumeRole"]
  }
}

resource "aws_iam_role" "test" {
  assume_role_policy = data.aws_iam_policy_document.trust.json
}

data "aws_iam_policy_document" "identity" {
  statement {
    effect = "Allow"

    actions = [
      "s3:*",
    ]
    resources = ["*"]
  }
}

resource "aws_iam_policy" "test" {
  policy = data.aws_iam_policy_document.identity.json
}

resource "aws_iam_role_policy_attachment" "test" {
  role       = aws_iam_role.test.name
  policy_arn = aws_iam_policy.test.arn
}
```

</details>

### Initial Reproduction

This section uses the same Terraform (`v1.0.5`) and AWS Provider (`v3.59.0`) versions provided in the original issue body.

```console
$ terraform -v
Terraform v1.0.5
on darwin_arm64
+ provider registry.terraform.io/hashicorp/aws v3.59.0
```

<details>
<summary>Show Commands</summary>

```console
$ terraform apply -auto-approve
<snip>
Plan: 3 to add, 0 to change, 0 to destroy.
aws_iam_role.test: Creating...
aws_iam_policy.test: Creating...
aws_iam_policy.test: Creation complete after 0s [id=arn:aws:iam::<redacted>:policy/terraform-20230215160339513800000001]
aws_iam_role.test: Creation complete after 0s [id=terraform-20230215160339514000000002]
aws_iam_role_policy_attachment.test: Creating...
aws_iam_role_policy_attachment.test: Creation complete after 0s [id=terraform-20230215160339514000000002-20230215160340132800000003]

Apply complete! Resources: 3 added, 0 changed, 0 destroyed.
```

```console
$ terraform plan
aws_iam_policy.test: Refreshing state... [id=arn:aws:iam::<redacted>:policy/terraform-20230215160339513800000001]
aws_iam_role.test: Refreshing state... [id=terraform-20230215160339514000000002]
aws_iam_role_policy_attachment.test: Refreshing state... [id=terraform-20230215160339514000000002-20230215160340132800000003]

Note: Objects have changed outside of Terraform

Terraform detected the following changes made outside of Terraform since the last "terraform apply":

  # aws_iam_policy.test has been changed
  ~ resource "aws_iam_policy" "test" {
        id        = "arn:aws:iam::<redacted>:policy/terraform-20230215160339513800000001"
        name      = "terraform-20230215160339513800000001"
      + tags      = {}
        # (5 unchanged attributes hidden)
    }
  # aws_iam_role.test has been changed
  ~ resource "aws_iam_role" "test" {
        id                    = "terraform-20230215160339514000000002"
      ~ managed_policy_arns   = [
          + "arn:aws:iam::<redacted>:policy/terraform-20230215160339513800000001",
        ]
        name                  = "terraform-20230215160339514000000002"
      + tags                  = {}
        # (9 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }

Unless you have made equivalent changes to your configuration, or ignored the relevant attributes using ignore_changes, the following plan may include actions to undo or respond to these changes.

──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

No changes. Your infrastructure matches the configuration.

Your configuration already matches the changes detected above. If you'd like to update the Terraform state to match, create and apply a refresh-only plan:
  terraform apply -refresh-only
```

</details>

As expected, the originally reported drift is observed. 

### Updating Terraform

Now, with the same state preserved, updating to the latest version of Terraform (`v1.3.8`) results in `No changes.`:

<details>
<summary>Show Commands</summary>

```console
$ terraform -v
Terraform v1.3.8
on darwin_arm64
+ provider registry.terraform.io/hashicorp/aws v3.59.0
```

```console
$ terraform plan
data.aws_caller_identity.current: Reading...
data.aws_iam_policy_document.identity: Reading...
data.aws_iam_policy_document.identity: Read complete after 0s [id=3228841632]
aws_iam_policy.test: Refreshing state... [id=arn:aws:iam::<redacted>:policy/terraform-20230215160339513800000001]
data.aws_caller_identity.current: Read complete after 0s [id=<redacted>]
data.aws_iam_policy_document.trust: Reading...
data.aws_iam_policy_document.trust: Read complete after 0s [id=759143916]
aws_iam_role.test: Refreshing state... [id=terraform-20230215160339514000000002]
aws_iam_role_policy_attachment.test: Refreshing state... [id=terraform-20230215160339514000000002-20230215160340132800000003]

No changes. Your infrastructure matches the configuration.

Terraform has compared your real infrastructure against your configuration and found no differences, so no changes are needed.
```

</details>

Switching between versions a bit more, I was able to determine the "Objects have changed outside of Terraform" note last appears with version `v1.1.9` (ie. was fixed in `v1.2.0`).

<details>
<summary>Show Commands</summary>

```console
$ terraform -v
Terraform v1.1.9
on darwin_arm64
+ provider registry.terraform.io/hashicorp/aws v3.59.0
```

```console
$ terraform plan
aws_iam_policy.test: Refreshing state... [id=arn:aws:iam::<redacted>:policy/terraform-20230215160339513800000001]
aws_iam_role.test: Refreshing state... [id=terraform-20230215160339514000000002]
aws_iam_role_policy_attachment.test: Refreshing state... [id=terraform-20230215160339514000000002-20230215160340132800000003]

Note: Objects have changed outside of Terraform

Terraform detected the following changes made outside of Terraform since the last "terraform apply":

  # aws_iam_policy.test has changed
  ~ resource "aws_iam_policy" "test" {
        id        = "arn:aws:iam::<redacted>:policy/terraform-20230215160339513800000001"
        name      = "terraform-20230215160339513800000001"
      + tags      = {}
        # (5 unchanged attributes hidden)
    }

  # aws_iam_role.test has changed
  ~ resource "aws_iam_role" "test" {
        id                    = "terraform-20230215160339514000000002"
      ~ managed_policy_arns   = [
          + "arn:aws:iam::<redacted>:policy/terraform-20230215160339513800000001",
        ]
        name                  = "terraform-20230215160339514000000002"
      + tags                  = {}
        # (9 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }


Unless you have made equivalent changes to your configuration, or ignored the relevant attributes using ignore_changes, the following plan may include actions to undo or respond to these changes.

──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

No changes. Your infrastructure matches the configuration.

Your configuration already matches the changes detected above. If you'd like to update the Terraform state to match, create and apply a refresh-only plan:
  terraform apply -refresh-only
```

```console
$ terraform -v
Terraform v1.2.0
on darwin_arm64
+ provider registry.terraform.io/hashicorp/aws v3.59.0
```

```console
$ terraform plan
data.aws_caller_identity.current: Reading...
data.aws_iam_policy_document.identity: Reading...
data.aws_iam_policy_document.identity: Read complete after 0s [id=3228841632]
aws_iam_policy.test: Refreshing state... [id=arn:aws:iam::<redacted>:policy/terraform-20230215160339513800000001]
data.aws_caller_identity.current: Read complete after 0s [id=<redacted>]
data.aws_iam_policy_document.trust: Reading...
data.aws_iam_policy_document.trust: Read complete after 0s [id=759143916]
aws_iam_role.test: Refreshing state... [id=terraform-20230215160339514000000002]
aws_iam_role_policy_attachment.test: Refreshing state... [id=terraform-20230215160339514000000002-20230215160340132800000003]

No changes. Your infrastructure matches the configuration.

Terraform has compared your real infrastructure against your configuration and found no differences, so no changes are needed.
```

</details>

**Comment by jar-b** [comment]
With no confirmed ongoing issues, we are closing this via https://github.com/hashicorp/terraform/pull/30486.
